cpp/monitoring/latency.h

```diff
@@ -0,0 +1,137 @@
+// duration between the call to ScopedLatency() and the destruction of the
+// returned object.
+//
+// The |TimeUnit| template parameter is used to specify the unit of the values
+// added to the counter, e.g. specifying std::chrono::milliseconds will
+// duration_cast all recorded latencies to milliseconds before adding to the
+// total count.
+//
+// |LabelTypes...| works as in the Counter<> and Gauge<> templates.
+//
+// Example usage:
+//
+//   static Latency<std::chrono::milliseconds, std::string> latency_by_name(
+//      "latency_by_name", "name");
+//   ...
+//
+//   void DoStuffForName(const string& name) {
+//     // This will record the latency of this method for each different |name|
+//     ScopedLatency latency(latency_by_name.ScopedLatency(name));
+//     ...
+//     // do stuff
+//     ...
+//   }
+//
+//   // Here's an example where there's no one scope so can't easily use the
+//   // ScopedLatency helper.
+//   void FinishedDoingStuffCallback(const string& name,
+//                                   const system_clock::time_point started_at)
+//                                   {
+//      latency_by_name.RecordLatency(name, system_clock::now() - started_at);
+//   }
+template <class TimeUnit, class... LabelTypes>
+class Latency {
+ public:
+  Latency(const std::string& base_name,
+          const typename NameType<LabelTypes>::name&... label_names,
+          const std::string& help);
+
+  void RecordLatency(const LabelTypes&... labels,
+                     std::chrono::duration<double> latency);
+
+  ScopedLatency ScopedLatency(const LabelTypes&... labels);
+
+ private:
+  std::mutex mutex_;
+  std::unique_ptr<Counter<LabelTypes...>> latencies_;
+  std::unique_ptr<Counter<LabelTypes...>> counts_;
+
+  DISALLOW_COPY_AND_ASSIGN(Latency);
+};
+
+
+// Helper class to automatically calculate and record latency.
+// Measures the duration between its construction and destruction times, and
+// automatically registers that with the Latency<> class which created it.
+class ScopedLatency {
+ public:
+  ScopedLatency(ScopedLatency&& other) = default;
+
+  ~ScopedLatency() {
+    record_latency_(std::chrono::system_clock::now() - start_);
+  }
+
+ private:
+  ScopedLatency(
+      const std::function<void(std::chrono::duration<double>)>& record_latency)
+      : record_latency_(record_latency),
+        start_(std::chrono::system_clock::now()) {
+  }
+
+  const std::function<void(std::chrono::duration<double>)> record_latency_;
+  const std::chrono::system_clock::time_point start_;
+
+  template <class TimeUnit, class... LabelTypes>
+  friend class Latency;
+
+  DISALLOW_COPY_AND_ASSIGN(ScopedLatency);
+};
+
+
+template <class TimeUnit, class... LabelTypes>
+Latency<TimeUnit, LabelTypes...>::Latency(
+    const std::string& base_name,
+    const typename NameType<LabelTypes>::name&... label_names,
+    const std::string& help)
+    : latencies_(Counter<LabelTypes...>::New(base_name, label_names..., help)),
```

Comment:
Wonder if it wouldn't be a good idea to also have a suffix here (say, "_overall_sum", for old time's sake), so that a computed variable built on that could have that name? Not sure how much sense that makes with Prometheus, though...